js/Integer.js

```diff
@@ -0,0 +1,74 @@
+import { InvalidInput } from "./Errors"
+
+
+export class Integer {
+    MIN = 1 << 32
+
+    static plus(a, b) {
+        Integer.#validate(a, b)
+
+        let carry = 0
+        while (b !== 0) {
+            carry = a & b
+            a = a ^ b
+            b = carry << 1
+        }
+
+        return a
+    }
+
+    static product(a, b) {
+        Integer.#validate(a, b)
+        let isNegativeResult = false
+        if (Integer.#isNegative(a)) {
+            a = Integer.negate(a)
+            isNegativeResult = !isNegativeResult
+        }
+        if (Integer.#isNegative(b)) {
+            b = Integer.negate(b)
+            isNegativeResult = !isNegativeResult
+        }
+
+        let result = 0
+        while (a !== 0) {
+            if (Integer.#isEven(a)) {
+                result = Integer.plus(result, b)
+            }
+            a = a >>> 1
+            b = b << 1
+        }
+
+        return isNegativeResult ? Integer.negate(result) : result
+    }
+
+    static #isNegative(integer) {
+        // For integers in Two's Complement, the 1st bit signifies whether the integer is negative.
+        // Thus, a bitwise AND operation with 1000... is equivalent to checking negativity.
+        return (integer & Integer.MIN) !== 0
+    }
+
+    static #isEven(integer) {
+        return (integer & 1) === 1
+    }
+
+    static negate(integer) {
+        return Integer.plus(Integer.#flipBits(integer), 1)
+    }
+
+    static #flipBits(integer) {
+        // An XOR comparison with 1 gives the opposite value of the other bit.
+        // The smallest Integer in Two's Complement consists of only 1's in binary representation.
```

Comment:
That is incorrect. Integer.MIN is 100..., what you're using is Integer.MIN >> 31.

I would also mention that this is all done to flip the bits of the input parameter. It's implicit from the method name, but spelling it out explicitly makes for an easier reading. Somehting like:
```JavaScript
// Flipping the bits of an integer is equal to an XOR comparison with 1 for every bit of the input.
// In Two's Complement, the number where all bits are equal to 1 can be generated by right shifting Integer.MIN (where only the most significant bit is set) until the leading 1 is shifted to the rightmost position.
```